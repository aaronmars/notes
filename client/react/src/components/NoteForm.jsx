import React from 'react';
import PropTypes from 'prop-types';

/**
 * A generic form component for editing a note.
 * This can be used for either creating or updating a note.
 */
export default class NoteForm extends React.Component {
  /**
   * Create a new NoteForm component
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    const { title, text } = props;
    this.state = { title, text };
    this.updateTitleInput = this.updateTitleInput.bind(this);
    this.updateTextInput = this.updateTextInput.bind(this);
    this.submit = this.submit.bind(this);
  }

  /**
   * Handler for when the title input value changes.
   * This keeps the state in sync with the form.
   * @param {React.SyntheticEvent} e
   */
  updateTitleInput(e) {
    const { onFieldChanged } = this.props;
    this.setState({ title: e.target.value });
    onFieldChanged('title', e.target.value);
  }

  /**
   * Handler for when the text input value changes.
   * This keeps the state in sync with the form.
   * @param {React.SyntheticEvent} e
   */
  updateTextInput(e) {
    const { onFieldChanged } = this.props;
    this.setState({ text: e.target.value });
    onFieldChanged('text', e.target.value);
  }

  /**
   * Handler for when the form is submitted.
   */
  submit() {
    const { onSubmit, operation } = this.props;
    const { title, text } = this.state;
    onSubmit(title, text);
    if (operation === 'add') {
      this.setState({ title: '', text: '' });
    }
  }

  /**
   * @return {React.Component<NoteForm>}
   */
  render() {
    const { operation, onCancel } = this.props;
    const { title, text } = this.state;
    return (
      <div>
        <div>
          Title:
          {' '}
          <input onChange={this.updateTitleInput} value={title} />
        </div>
        <div>
          Text:
          {' '}
          <input onChange={this.updateTextInput} value={text} />
        </div>
        <button onClick={this.submit} type="submit">
          {operation === 'add' ? 'Add Note' : 'Save Note'}
        </button>
        {operation === 'update' && <button onClick={onCancel} type="submit">Cancel</button>}
      </div>
    );
  }
}
NoteForm.propTypes = {
  title: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  operation: PropTypes.string.isRequired,
  onFieldChanged: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func,
};
NoteForm.defaultProps = {
  onCancel: () => {},
};
