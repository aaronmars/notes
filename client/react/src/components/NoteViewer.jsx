import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchNote } from '../../../common/actions';

/**
 * A simple viewer for a note based on the `currentNote` field from the store.
 */
class NoteViewer extends React.Component {
  /**
   * Create a new NoteViewer component,
   * It can be populated on construction by supplying
   *   a `match` object from a router.
   * @param {*} props
   */
  constructor(props) {
    super(props);
    const { match, fetchNote: dispatchFetchNote } = props;
    if (match) {
      dispatchFetchNote(match.params.noteId, true);
    }
  }

  /**
   * @return {React.Component<NoteEditor>}
   */
  render() {
    const { currentNote } = this.props;
    return currentNote === null ? (
      <p>No note selected</p>
    ) : (
      <div>
        <p>
          Now viewing note ID
          {currentNote.id}
        </p>
        <h1>{currentNote.title}</h1>
        <h2>{currentNote.text}</h2>
      </div>
    );
  }
}
NoteViewer.propTypes = {
  match: PropTypes.shape({ params: PropTypes.shape({ noteId: PropTypes.string.isRequired }) }),
  fetchNote: PropTypes.func.isRequired,
  currentNote: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }),
};
NoteViewer.defaultProps = {
  match: { params: { noteId: null } },
  currentNote: null,
};

const mapStateToProps = state => ({ currentNote: state.currentNote });

export default connect(
  mapStateToProps,
  { fetchNote },
)(NoteViewer);
