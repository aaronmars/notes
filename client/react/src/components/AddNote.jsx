import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import NoteForm from './NoteForm';
import { addNote } from '../../../common/actions';

/**
 * A component for adding a new note.
 */
class AddNote extends React.Component {
  /**
   * Create a new AddNote component
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    this.state = { title: '', text: '' };
    this.addNewNote = this.addNewNote.bind(this);
    this.fieldChanged = this.fieldChanged.bind(this);
  }

  /**
   * Adds a note to the store, and then resets the component state
   */
  addNewNote() {
    const { title, text } = this.state;
    const { addNote: dispatchAddNote } = this.props;
    dispatchAddNote(title, text).then(() => {
      this.setState({ title: '', text: '' });
    });
  }

  /**
   * Updates the state when the title or text fields change
   * @param {*} field
   * @param {*} value
   */
  fieldChanged(field, value) {
    this.setState({ [field]: value });
  }

  /**
   * @return {React.Component<AddNote>} A form for adding a note.
   */
  render() {
    const { title, text } = this.state;
    return (
      <NoteForm
        title={title}
        text={text}
        operation="add"
        onSubmit={this.addNewNote}
        onFieldChanged={this.fieldChanged}
      />
    );
  }
}
AddNote.propTypes = {
  addNote: PropTypes.func.isRequired,
};

export default connect(
  null,
  { addNote },
)(AddNote);
