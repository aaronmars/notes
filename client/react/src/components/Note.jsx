import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import NoteForm from './NoteForm';
import { updateNote, fetchNote, selectNote } from '../../../common/actions';

/**
 * A component for a single note within a note listing.
 * This component contains functionality to update the note as well.
 */
class Note extends React.Component {
  /**
   * Create a new Note component
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    const { title, text } = props;
    this.state = {
      mode: 'view',
      editTitle: title,
      editText: text,
    };
    this.editClicked = this.editClicked.bind(this);
    this.fieldChanged = this.fieldChanged.bind(this);
    this.updateNoteCancelled = this.updateNoteCancelled.bind(this);
    this.updateNote = this.updateNote.bind(this);
    this.selectNote = this.selectNote.bind(this);
  }

  /**
   * Handler for this component's edit button is clicked.
   */
  editClicked() {
    this.setState({ mode: 'edit' });
  }

  /**
   * Handler for when one of the fields in the editor form changes.
   * This sets the state so the temporary changes to the fields are
   *   available in case the update is submitted.
   * @param {*} field
   * @param {*} value
   */
  fieldChanged(field, value) {
    if (field === 'title') {
      this.setState({ editTitle: value });
    } else if (field === 'text') {
      this.setState({ editText: value });
    }
  }

  /**
   * Handler for when the update operation is cancelled.
   * This sets the component back to view mode.
   */
  updateNoteCancelled() {
    this.setState({ mode: 'view' });
  }

  /**
   * Handler for when the update operation is submitted.
   * This persists the update to the store.
   */
  updateNote() {
    const {
      updateNote: dispatchUpdateNote,
      selectNote: dispatchSelectNote,
      id,
      currentNote,
    } = this.props;
    const { editTitle, editText } = this.state;
    dispatchUpdateNote(id, editTitle, editText).then(() => {
      // Now that the update operation is done, switch to view mode
      //  and update the current note in case is is this note
      this.setState({ mode: 'view' });
      if (currentNote.id === id) {
        dispatchSelectNote(id, editTitle, editText);
      }
    });
  }

  /**
   * Handler for when the note is selected.
   */
  selectNote() {
    const { fetchNote: dispatchFetchNote, id } = this.props;
    dispatchFetchNote(id);
  }

  /**
   * @return {React.Component} The Note component
   */
  render() {
    const { id, title, text } = this.props;
    const { mode } = this.state;
    return mode === 'view' ? (
      <div>
        <div>
          <h3>{title}</h3>
        </div>
        <div>{text}</div>
        <button onClick={this.editClicked} type="submit">Edit</button>
        <Link to={`/${id}`} onClick={this.selectNote}>
          Select
        </Link>
      </div>
    ) : (
      <NoteForm
        operation="update"
        title={title}
        text={text}
        onSubmit={this.updateNote}
        onCancel={this.updateNoteCancelled}
        onFieldChanged={this.fieldChanged}
      />
    );
  }
}
Note.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  updateNote: PropTypes.func.isRequired,
  selectNote: PropTypes.func.isRequired,
  fetchNote: PropTypes.func.isRequired,
  currentNote: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }),
};
Note.defaultProps = {
  currentNote: null,
};

const mapStateToProps = state => ({
  currentNote: state.currentNote,
});

export default connect(
  mapStateToProps,
  { updateNote, fetchNote, selectNote },
)(Note);
