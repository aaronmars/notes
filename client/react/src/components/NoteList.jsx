import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Note from './Note';

/**
 * Render a list of Notes fetched from the API
 */
export function NoteList(props) {
  const { notes } = props;
  return (
    <div>
      {Object.entries(notes).map(([id, note]) => (
        <Note key={id} title={note.title} text={note.text} id={id} />
      ))}
    </div>
  );
}
NoteList.propTypes = {
  notes: PropTypes.shape({}).isRequired,
};

const mapStateToProps = state => ({ notes: state.notes });

export default connect(mapStateToProps)(NoteList);
