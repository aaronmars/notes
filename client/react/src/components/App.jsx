import React from 'react';
import { BrowserRouter, Route } from 'react-router-dom';
import AddNote from './AddNote';
import NoteList from './NoteList';
import NoteViewer from './NoteViewer';

/**
 * The main app.
 * This features a note listing, and a note viewer.
 * The note viewer is directed when the final segment in the URL is a note ID
 */
export default class App extends React.Component {
  /**
   * Create a new App component
   * @param {Object} props
   */
  constructor(props) {
    super(props);

    // Here we have the only styling in the whole app!
    // These are just here to make the listing
    //   and the viewer appear side-by-side
    this.flexStyle = { display: 'flex' };
    this.containerStyle = { padding: '1em' };
  }

  /**
   * @return {React.Component<App>} The app component
   */
  render() {
    return (
      <BrowserRouter basename="/react">

        <div style={this.flexStyle}>
          <div style={this.containerStyle}>
            <AddNote />
            <NoteList />
          </div>

          <div style={this.containerStyle}>
            <Route path="/:noteId" component={NoteViewer} />
          </div>
        </div>

      </BrowserRouter>
    );
  }
}
