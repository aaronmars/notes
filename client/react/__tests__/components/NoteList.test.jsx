import React from 'react';
import renderer from 'react-test-renderer';
import { NoteList } from '../../src/components/NoteList';

jest.mock('../../src/components/Note', () => 'Note');
test('Empty NoteList is rendered', () => {
  const component = renderer.create(
    <NoteList notes={{}} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Populated NoteList is rendered', () => {
  const notesData = {
    id1: { title: 'Note1', text: 'Note1 text' },
    id2: { title: 'Note2', text: 'Note2 text' },
  };
  const component = renderer.create(<NoteList notes={notesData} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
