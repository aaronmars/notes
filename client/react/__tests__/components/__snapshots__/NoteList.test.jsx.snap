// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty NoteList is rendered 1`] = `<div />`;

exports[`Populated NoteList is rendered 1`] = `
<div>
  <Note
    id="id1"
    text="Note1 text"
    title="Note1"
  />
  <Note
    id="id2"
    text="Note2 text"
    title="Note2"
  />
</div>
`;
